### PR TITLE
[AMDGPU] Reland: Do not use original PHIs in coercion chains

### PR DESCRIPTION
**Pull request by jrbyrnes**
Reland https://github.com/llvm/llvm-project/commit/1612e4a3510982692f22e3f8190fc7c977185cbe

The original version didn't correctly remove the broken PHI chain from the ValMap, so it would still be used in some cases. 

During this, I spotted another issue where users try to use the type coerced version of the value even though the original value was defined in the same block. I've also addressed that in this reland.

### CODE REVIEW

File: llvm/lib/Target/AMDGPU/AMDGPULateCodeGenPrepare.cpp [hunk 1]
```diff
@@ -384,10 +404,18 @@ bool LiveRegOptimizer::optimizeLiveType(
           NewVal = BBUseValMap[U->getParent()][ValMap[Op]];
         else {
           BasicBlock::iterator InsertPt = U->getParent()->getFirstNonPHIIt();
-          NewVal =
-              convertFromOptType(Op->getType(), cast<Instruction>(ValMap[Op]),
-                                 InsertPt, U->getParent());
-          BBUseValMap[U->getParent()][ValMap[Op]] = NewVal;
+          // We may pick up ops that were previously converted for users in
+          // other blocks. If there is a originally typed definition of the Op
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/AMDGPULateCodeGenPrepare.cpp:408]:
```suggestion
          // other blocks. If there is an originally typed definition of the Op
```

File: llvm/lib/Target/AMDGPU/AMDGPULateCodeGenPrepare.cpp [hunk 2]
```diff
@@ -365,13 +365,33 @@ bool LiveRegOptimizer::optimizeLiveType(
...
         MissingIncVal = true;
     }
-    Instruction *DeadInst = Phi;
     if (MissingIncVal) {
-      DeadInst = cast<Instruction>(ValMap[Phi]);
-      // Do not use the dead phi
-      ValMap[Phi] = Phi;
+      Value *DeadVal = ValMap[Phi];
+      // The coercion chain of the PHI is broken. Delete the Phi
+      // from the ValMap and any connected / user Phis.
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/AMDGPULateCodeGenPrepare.cpp:371]: Isn't this what ValueMap is supposed to take care of for you? The ValueHandles should have been updated when you replace the value?

Comment (jrbyrnes) [reply llvm/lib/Target/AMDGPU/AMDGPULateCodeGenPrepare.cpp:371]: Not quite sure how we could more easily accomplish this using ValueHandles -- We never RAUW so by simply replacing to use ValueHandles we would still be where we are. 

But if we changed the implementation of PHI node handling to use RAUW, it seems to me that we would still create broken PHIs and would need some unwinding mechanism.

**Comment by jrbyrnes** [comment]
Ping

After revert, staging breaks on rocPRIM.

Passes psdb +  Fixes: SWDEV-474267 SWDEV-473447 SWDEV-472000

Is it possible to refactor for ValueHandles in future iteration?

File: llvm/lib/Target/AMDGPU/AMDGPULateCodeGenPrepare.cpp (hunk 2 above)
Comment (arsenm) [reply llvm/lib/Target/AMDGPU/AMDGPULateCodeGenPrepare.cpp:371]: But if you do RAUW, the weak handles handle the case where the value is ultimately deleted

**Review by arsenm — APPROVED** [review]
LGTM but I think this should try to use ValueMap and/or WeakTrackingVH and replacements. This approach seems more complicated than necessary to keep values updated

**Comment by jrbyrnes** [comment]
https://github.com/llvm/llvm-project/commit/6e68b75e6634e9809e2af5d68c87a4eb4282217e

Thanks for review again. I agree that this approach is likely not the best, and will find a better way to deal with broken chains.

**Pull request closed without merging**